#!/usr/bin/env bash

##############################################################################################################################################################
# Author of pipeline: Haitao Han.
# For questions, bugs, and suggestions, contact me at htaohan@163.com
##############################################################################################################################################################
VERSION="1.0.0"

help_message () {
  echo ""
  echo "databinning version: $VERSION"
	echo "Options:"
	echo ""
	echo "  -a STR          metagenomic assembly file"
	echo "  -o STR          output directory"
	echo "  -p STR          path to access to the bam files"
	echo "  -m STR          metagenomic binning method"
	echo "  -t INT          number of threads"
	echo "  -s STR          sam files"
	echo "  -b STR          bam files"
	echo "  -h STR          depth file generated by MetaBAT 2"
	echo "  -x STR          Path to the bins that need to be refined"
	echo "  -y STR          Path to the bins that need to be refined"
	echo "  -z STR          Path to the bins that need to be refined"
	echo "";}

if [[ $1 == "-h" ]]; then
    help_message
    exit 0
fi

while getopts a:o:p:b:m:t:s:f:x:y:z: OPT; do
 case ${OPT} in
  a) contig_file=$(realpath ${OPTARG})
    ;;
  o) output_dir=$(realpath ${OPTARG})
    ;;
  p) bam_file_path=$(realpath ${OPTARG})
    ;;
  b) bam_files=$(realpath ${OPTARG})
    ;;
  m) binner=${OPTARG}
    ;;
  t) threads=${OPTARG}
    ;;
  s) sam_files=$(realpath ${OPTARG})
    ;;
  f) depth_file=$(realpath ${OPTARG})
    ;;
  x) bin1_dir=$(realpath ${OPTARG})
    ;;
  y) bin2_dir=$(realpath ${OPTARG})
    ;;
  z) bin3_dir=$(realpath ${OPTARG})
    ;;
  \?)
    exit 1
 esac
done



case ${binner} in
    metabat2)
#        if [ -z "${contig_file}" ] || [ -z "${output_dir}" ]; then
#            echo "Error: For mode 'metabat', both -a <file> and -b <dir> are required."
#            help_message
#        fi
        echo "Executing MetaBat 2 with threads ${threads} contigfile ${contig_file} -bamfile ${bam_files}"
        mkdir -p ${output_dir}/bins_dir
        cd ${output_dir}
        jgi_summarize_bam_contig_depths --outputDepth depth.txt ${bam_files}
        metabat2 -t ${threads} -i ${contig_file} -a depth.txt -o ${output_dir}/bins_dir/metabat2
        ;;
    concoct)
#        if [ -z "${contig_file}" ] || [ -z "${output_dir}" ]; then
#            echo "Error: For mode 'metabat', both -a <file> and -b <dir> are required."
#            help_message
        echo "Executing Concoct with threads ${threads} contigfile ${contig_file} -bamfile ${bam_files}"
        mkdir -p ${output_dir}/bins_dir
        cd ${output_dir}
        cut_up_fasta.py ${contig_file} -c 10000 -o 0 --merge_last -b contigs_10K.bed > contigs_10K.fa
        concoct_coverage_table.py contigs_10K.bed ${bam_files} > coverage_table.tsv
        concoct --composition_file contigs_10K.fa --coverage_file coverage_table.tsv --threads ${threads} -o bins_dir/concoct
        ;;
    comebin)
#        if [ -z "${contig_file}" ] || [ -z "${output_dir}" ]; then
#            echo "Error: For mode 'metabat', both -a <file> and -b <dir> are required."
#            help_message
#        fi
        echo "Executing COMEBin with threads ${threads} contigfile ${contig_file} bamfilepath ${bam_file_path}"
        mkdir -p ${output_dir}
        cd ${output_dir}
        run_comebin.sh -t ${threads} -a ${contig_file} -o ${output_dir} -p ${bam_file_path}
        cp -r ${output_dir}/comebin_res/comebin_res_bins ${output_dir}/bins_dir
        ;;
    metadecoder)
#        if [ -z "${contig_file}" ] || [ -z "${output_dir}" ]; then
#            echo "Error: For mode 'metabat', both -a <file> and -b <dir> are required."
#            help_message
#        fi
        echo "Executing MetaDecoder with threads ${threads} contigfile ${contig_file} samfile ${sam_files}"
        mkdir -p ${output_dir}/bins_dir
        cd ${output_dir}
        metadecoder coverage --threads ${threads} -s ${sam_files}  -o METADECODER_gsa.COVERAGE
        metadecoder seed --threads ${threads} -f ${contig_file} -o METADECODER_gsa.SEED
        metadecoder cluster -f ${contig_file} -c METADECODER_gsa.COVERAGE -s METADECODER_gsa.SEED -o bins_dir/METADECODER
        ;;
    metabinner)
#        if [ -z "${contig_file}" ] || [ -z "${output_dir}" ]; then
#            echo "Error: For mode 'metabat', both -a <file> and -b <dir> are required."
#            help_message
#        fi
        echo "Executing MetaBinner with threads ${threads} contigfile ${contig_file} depthfile ${depth_file}"
        mkdir -p ${output_dir}
        awk -F'\t' '{OFS="\t"; output=""; for(i=1; i<=NF; i++) if (i == 1 || (i >= 4 && (i-4) % 2 == 0)) { if (output != "") output = output OFS; output = output $i } print output}' ${depth_file} > ${output_dir}/coverage_profile.tsv
        cd $CONDA_PREFIX/bin/MetaBinner/scripts
        metabinner_path=$CONDA_PREFIX/bin/MetaBinner
        python gen_kmer.py ${contig_file} 1000 4 ${output_dir}/kmer.tsv

        ../run_metabinner.sh -t ${threads} -a ${contig_file} -o ${output_dir} -d ${output_dir}/coverage_profile.tsv -k ${output_dir}/kmer.tsv -p ${metabinner_path}
        cp -r ${output_dir}/metabinner_res/ensemble_res/greedy_cont_weight_3_mincomp_50.0_maxcont_15.0_bins/ensemble_3logtrans/addrefined2and3comps/greedy_cont_weight_3_mincomp_50.0_maxcont_15.0_bins ${output_dir}/bins_dir
        ;;
    magscot)
#        if [ -z "${contig_file}" ] || [ -z "${output_dir}" ]; then
#            echo "Error: For mode 'metabat', both -a <file> and -b <dir> are required."
#            help_message
#        fi
        echo "Executing MAGScoT with threads ${threads} contigfile ${contig_file} "
        MAGScoT_folder=${CONDA_PREFIX}/bin/MAGScoT

        cd ${output_dir}
        gzip -c ${contig_file} > example.contigs.fasta.gz
        mkdir -p tmp_workfolder
        zcat example.contigs.fasta.gz | parallel -j ${threads} --block 999k --recstart '>' --pipe prodigal -p meta -a tmp_workfolder/example.{\#}.faa -d tmp_workfolder/example.{\#}.ffn -o tmpfile
        cat tmp_workfolder/example.*.faa > example.prodigal.faa
        cat tmp_workfolder/example.*.ffn > example.prodigal.ffn
        rm -r tmp_workfolder tmpfile

        hmmsearch -o example.hmm.tigr.out --tblout example.hmm.tigr.hit.out --noali --notextw --cut_nc --cpu ${threads} $MAGScoT_folder/hmm/gtdbtk_rel207_tigrfam.hmm example.prodigal.faa
        hmmsearch -o example.hmm.pfam.out --tblout example.hmm.pfam.hit.out --noali --notextw --cut_nc --cpu ${threads} $MAGScoT_folder/hmm/gtdbtk_rel207_Pfam-A.hmm example.prodigal.faa

        cat example.hmm.tigr.hit.out | grep -v "^#" | awk '{print $1"\t"$3"\t"$5}' > example.tigr
        cat example.hmm.pfam.hit.out | grep -v "^#" | awk '{print $1"\t"$4"\t"$5}' > example.pfam
        cat example.pfam example.tigr > example.hmm

        Contigs_to_bin_tsv.py --path ${bin1_dir} -o ${bin1_dir}/contig_to_bins.tsv
        Contigs_to_bin_tsv.py --path ${bin2_dir} -o ${bin2_dir}/contig_to_bins.tsv
        Contigs_to_bin_tsv.py --path ${bin3_dir} -o ${bin3_dir}/contig_to_bins.tsv
        chmod -R 777 ${bin1_dir}/contig_to_bins.tsv
        chmod -R 777 ${bin2_dir}/contig_to_bins.tsv
        chmod -R 777 ${bin3_dir}/contig_to_bins.tsv

        awk '{print $2"\t"$1"\tBinner1"}'  ${bin1_dir}/contig_to_bins.tsv > example.contigs_to_bin.tsv
        awk '{print $2"\t"$1"\tBinner2"}'  ${bin2_dir}/contig_to_bins.tsv >> example.contigs_to_bin.tsv
        awk '{print $2"\t"$1"\tBinner3"}'  ${bin3_dir}/contig_to_bins.tsv >> example.contigs_to_bin.tsv

        Rscript $MAGScoT_folder/MAGScoT.R -i example.contigs_to_bin.tsv --hmm example.hmm

        mkdir resulted_bins
        extract_fasta_bins.py ${contig_file} MAGScoT.refined.contig_to_bin.out --output_path resulted_bins
        ;;
    auto)
        ## run metabat 2
        echo "Executing MetaBat 2 with threads ${threads} contigfile ${contig_file} -bamfile ${bam_files}"
        mkdir -p ${output_dir}/metabat2_resul/bins_dir 
        cd ${output_dir}/metabat2_result
        jgi_summarize_bam_contig_depths --outputDepth depth.txt ${bam_files}
        mkdir bins_dir
        metabat2 -t ${threads} -i ${contig_file} -a depth.txt -o ${output_dir}/metabat2_resul/bins_dir 

        ## run concoct
        echo "Executing Concoct with threads ${threads} contigfile ${contig_file} -bamfile ${bam_files}"
        mkdir -p ${output_dir}/concoct_result/bins_dir
        cd ${output_dir}/concoct_result
        mkdir -p bins_dir
        cut_up_fasta.py ${contig_file} -c 10000 -o 0 --merge_last -b contigs_10K.bed > contigs_10K.fa
        concoct_coverage_table.py contigs_10K.bed ${bam_files} > coverage_table.tsv
        concoct --composition_file contigs_10K.fa --coverage_file coverage_table.tsv --threads ${threads} -o ${output_dir}/concoct_result/bins_dir
        
        ##run metadecoder
        echo "Executing MetaDecoder with threads ${threads} contigfile ${contig_file} samfile ${sam_files}"
        mkdir -p ${output_dir}/metadecoder_result/bins_dir
        cd ${output_dir}/metadecoder_result
        metadecoder coverage --threads ${threads} -s ${sam_files}  -o METADECODER_gsa.COVERAGE
        metadecoder seed --threads ${threads} -f ${contig_file} -o METADECODER_gsa.SEED
        metadecoder cluster -f ${contig_file} -c METADECODER_gsa.COVERAGE -s METADECODER_gsa.SEED -o ${output_dir}/metadecoder_result/bins_dir

        ##run metabinner
        echo "Executing MetaBinner with threads ${threads} contigfile ${contig_file} depthfile ${depth_file}"
        mkdir -p ${output_dir}/metabinner_result/bins_dir
        awk -F'\t' '{OFS="\t"; output=""; for(i=1; i<=NF; i++) if (i == 1 || (i >= 4 && (i-4) % 2 == 0)) { if (output != "") output = output OFS; output = output $i } print output}' ${output_dir}/metabat2_result/depth.txt > ${output_dir}/metabinner_result/coverage_profile.tsv
        cd $CONDA_PREFIX/bin/MetaBinner/scripts
        metabinner_path=$CONDA_PREFIX/bin/MetaBinner
        python gen_kmer.py ${contig_file} 1000 4 ${output_dir}/metabinner_result/kmer.tsv

        ../run_metabinner.sh -t ${threads} -a ${contig_file} -o ${output_dir}/metabinner_result -d ${output_dir}/metabinner_result/coverage_profile.tsv -k ${output_dir}/metabinner_result/kmer.tsv -p ${metabinner_path}
        cp -r ${output_dir}/metabinner_result/metabinner_res/ensemble_res/greedy_cont_weight_3_mincomp_50.0_maxcont_15.0_bins/ensemble_3logtrans/addrefined2and3comps/greedy_cont_weight_3_mincomp_50.0_maxcont_15.0_bins ${output_dir}/metabinner_result/bins_dir
        ## run comebin
        echo "Executing COMEBin with threads ${threads} contigfile ${contig_file} bamfilepath ${bam_file_path}"
        mkdir -p ${output_dir}/comebin_result/bins_dir
        cd ${output_dir}/comebin_result
        mkdir bamfiles_tmp
        cp ${bam_files} bamfiles_tmp
        run_comebin.sh -t ${threads} -a ${contig_file} -o ${output_dir}/comebin_result -p bamfiles_tmp
        cp -r ${output_dir}/comebin_result/comebin_res/comebin_res_bins ${output_dir}/comebin_result/bins_dir
        rm -r bamfiles_tmp
        ## refinement
        echo "Executing MAGScoT with threads ${threads} contigfile ${contig_file} "
        MAGScoT_folder=${CONDA_PREFIX}/bin/MAGScoT

        mkdir -p ${output_dir}/refined_result
        cd ${output_dir}/refined_result
        gzip -c ${contig_file} > example.contigs.fasta.gz
        mkdir -p tmp_workfolder
        zcat example.contigs.fasta.gz | parallel -j ${threads} --block 999k --recstart '>' --pipe prodigal -p meta -a tmp_workfolder/example.{\#}.faa -d tmp_workfolder/example.{\#}.ffn -o tmpfile
        cat tmp_workfolder/example.*.faa > example.prodigal.faa
        cat tmp_workfolder/example.*.ffn > example.prodigal.ffn
        rm -r tmp_workfolder tmpfile

        hmmsearch -o example.hmm.tigr.out --tblout example.hmm.tigr.hit.out --noali --notextw --cut_nc --cpu ${threads} $MAGScoT_folder/hmm/gtdbtk_rel207_tigrfam.hmm example.prodigal.faa
        hmmsearch -o example.hmm.pfam.out --tblout example.hmm.pfam.hit.out --noali --notextw --cut_nc --cpu ${threads} $MAGScoT_folder/hmm/gtdbtk_rel207_Pfam-A.hmm example.prodigal.faa

        cat example.hmm.tigr.hit.out | grep -v "^#" | awk '{print $1"\t"$3"\t"$5}' > example.tigr
        cat example.hmm.pfam.hit.out | grep -v "^#" | awk '{print $1"\t"$4"\t"$5}' > example.pfam
        cat example.pfam example.tigr > example.hmm

        Contigs_to_bin_tsv.py --path ${output_dir}/metabat2_result/bins_dir -o ${output_dir}/metabat2_result/contig_to_bins.tsv
        Contigs_to_bin_tsv.py --path ${output_dir}/concoct_result/bins_dir -o ${output_dir}/concoct_result/contig_to_bins.tsv
        Contigs_to_bin_tsv.py --path ${output_dir}/metabinner_result/bins_dir -o ${output_dir}/metabinner_result/contig_to_bins.tsv
        Contigs_to_bin_tsv.py --path ${output_dir}/comebin_result/bins_dir -o ${output_dir}/comebin_result/contig_to_bins.tsv
        Contigs_to_bin_tsv.py --path ${output_dir}/metadecoder_result/bins_dir -o ${output_dir}/metadecoder_result/contig_to_bins.tsv

        chmod -R 777 ${output_dir}/metabat2_result/contig_to_bins.tsv
        chmod -R 777 ${output_dir}/concoct_result/contig_to_bins.tsv
        chmod -R 777 ${output_dir}/metabinner_result/contig_to_bins.tsv
        chmod -R 777 ${output_dir}/comebin_result/contig_to_bins.tsv
        chmod -R 777 ${output_dir}/metadecoder_result/contig_to_bins.tsv

        awk '{print $2"\t"$1"\tconcoct"}'  ${output_dir}/concoct_result/contig_to_bins.tsv >> example.contigs_to_bin.tsv
        awk '{print $2"\t"$1"\tmetabat2"}'  ${output_dir}/metabat2_result/contig_to_bins.tsv > example.contigs_to_bin.tsv
        awk '{print $2"\t"$1"\tmetabinner"}'  ${output_dir}/metabinner_result/contig_to_bins.tsv >> example.contigs_to_bin.tsv
        awk '{print $2"\t"$1"\tcomebin"}'  ${output_dir}/comebin_result/contig_to_bins.tsv >> example.contigs_to_bin.tsv
        awk '{print $2"\t"$1"\tmetadecoder"}'  ${output_dir}/metadecoder_result/contig_to_bins.tsv >> example.contigs_to_bin.tsv

        Rscript $MAGScoT_folder/MAGScoT.R -i example.contigs_to_bin.tsv --hmm example.hmm

        mkdir bins_dir
        extract_fasta_bins.py ${contig_file} MAGScoT.refined.contig_to_bin.out --output_path bins_dir
        ;;
    *)
        echo "Error: Invalid binner '${binner}'."
        help_message
        ;;
esac